File: 2/2000s/24s/2444/CountSubarraysWithFixedBounds.js
/*
https://leetcode.com/problems/count-subarrays-with-fixed-bounds/description/?envType=daily-question&envId=2025-04-26
*/

/**
 * @param {number[]} nums
 * @param {number} minK
 * @param {number} maxK
 * @return {number}
 */
var countSubarrays = function(nums, minK, maxK) {
    let lengthOfNums=nums.length, lastMinIndex=-1, lastMaxIndex=-1, lastBadIndex=-1;
    let res=0;
    for(let i=0; i<lengthOfNums; ++i){
        if(nums[i]<minK||nums[i]>maxK){
            lastBadIndex=i;
        }
        if(nums[i]==minK) lastMinIndex=i;
        if(nums[i]==maxK) lastMaxIndex=i;
        let curValidIndex=Math.min(lastMinIndex, lastMaxIndex);
        if(curValidIndex>lastBadIndex){
            res+=(curValidIndex-lastBadIndex);
        }
    }
    return res;
};

File: 2/2000s/24s/2444/CountSubarraysWithFixedBounds.cpp
/*
https://leetcode.com/problems/count-subarrays-with-fixed-bounds/description/?envType=daily-question&envId=2025-04-26
*/

class Solution {
public:
    long long countSubarrays(vector<int>& nums, int minK, int maxK) {
        int sz=nums.size(), lastMinIdx=-1, lastMaxIdx=-1, lastBadIdx=-1;
        long long res=0;
        for(int i=0; i<sz; ++i){
            if(nums[i]<minK||nums[i]>maxK){
                lastBadIdx=i;
            }
            if(nums[i]==minK){
                lastMinIdx=i;
            }
            if(nums[i]==maxK){
                lastMaxIdx=i;
            }
            int curValidIdx=min(lastMinIdx, lastMaxIdx);
            if(curValidIdx>lastBadIdx){
                res+=(curValidIdx-lastBadIdx);
            }   
        }
        return res;
    }
};